### DISCUSSION TITLE
use of hardware timer to manually perform PWM for port manipulation

### DISCUSSION THREAD

**Discussion by KChegwin** [earlephilhower/arduino-pico#536] [General]
Hi all.
I am using a generic RP2040 and setting the analog write frequency to 20000 Hz with a 10 bit write resolution.
I'm intending to perform regenerative braking for a motor controller. This will require me to toggle multiple H bridge pins in sync. It would seem that port manipulation offers a means to digitally toggle pins simultaneously, however as I am performing motor control with PWM I am in search of an analogWrite equivalent. I was wondering if anyone has attempted to do this using the hardware timers on the RP2040?

thanks, KChegwin

**Comment by khoih-prog** [comment]
You can try my [**RP2040_PWM**](https://github.com/khoih-prog/RP2040_PWM) to see if OK for you.

> This PWM-wrapper library enables you to use Hardware-PWM blocks on RP2040-based boards to create and output PWM any GPIO pin. These purely hardware-based PWM channels can generate from very low (lowest is 7.5Hz) to very high PWM frequencies (in the MHz range, up to 62.5MHz).

**Reply by earlephilhower** [reply]
FWIW, the Arduino core already used the HW PWM module to generate the actual PWM, not timer interrupts.  Standard analogWrite/etc.

I'm not sure what exactly @KChegwin is looking for...if all the H bridge pins need to be driven at the same time, all the time, then just using 1 IO pin (and maybe an external 74XX buffer) is all you need.  If you need phase relationships between the pins, that's a little different and will need something fancier than what's in the core...

**Reply by KChegwin** [reply]
thanks for the suggestion @khoih-prog , I was able to get the library up and running but found that assigning different duty cycles to 2 separtate pins (using a modified version of the dynamic duty cycle example) caused their phase relationships to change when viewed through an oscilloscope. Perhaps as @earlephilhower suggests, this is beyond the capability of the core and requires some sort of alternate hardware solution.

thanks, appreciate the help
